infrastructure/src/main/java/org/corfudb/infrastructure/SequencerServerCache.java

```diff
@@ -62,159 +78,160 @@ public class SequencerServerCache implements AutoCloseable {
+    public long get(@NonNull ConflictTxStream conflictKey) {
+        return conflictKeyMap.getOrDefault(conflictKey, Address.NON_ADDRESS);
     }
 
     /**
-     * The first address in the priority queue.
+     * Returns the first (smallest) address in the cache.
+     * @return The smallest address currently in the cache.
      */
     public long firstAddress() {
-        if (cacheEntries.isEmpty()) {
+        if (versionMap.isEmpty()) {
             return Address.NOT_FOUND;
         }
-        return cacheEntries.peek().txVersion;
-    }
 
-  /**
-   * Invalidate the records with the minAddress. It could be one or multiple records
-   *
-   * @return the number of entries has been invalidated and removed from the cache.
-   */
-  private int invalidateSmallestTxVersion() {
-    ConflictTxStream firstEntry = cacheEntries.peek();
-    if (cacheEntries.size() == 0) {
-      return 0;
+        return versionMap.firstKey();
     }
 
-    int numEntries = 0;
-
-    while (firstAddress() == firstEntry.txVersion) {
-      if (log.isTraceEnabled()) {
-        log.trace(
-            "invalidateSmallestTxVersion: items evicted {} min address {}",
-            numEntries,
-            firstAddress());
-      }
-      ConflictTxStream entry = cacheEntries.poll();
-      conflictKeys.remove(entry);
-      numEntries++;
+    /**
+     * Evict the records with the smallest address in the cache.
+     * It could be one or multiple records.
+     *
+     * @return The number of entries that have been evicted from the cache.
+     */
+    private int evictSmallestTxVersion() {
+      final long firstAddress = firstAddress();
+      final Set<ConflictTxStream> entriesToDelete = versionMap.remove(firstAddress);
+      final int numDeletedEntries = entriesToDelete.size();
+      log.trace("evictSmallestTxVersion: items evicted {} min address {}", numDeletedEntries, firstAddress);
+      entriesToDelete.forEach(conflictKeyMap::remove);
+      maxConflictWildcard = Math.max(maxConflictWildcard, firstAddress);
+      return numDeletedEntries;
     }
 
-    maxConflictWildcard = Math.max(maxConflictWildcard, firstEntry.txVersion);
-    return numEntries;
-  }
-
     /**
-     * Invalidate all records up to a trim mark (not included).
-     *
-     * @param trimMark trim mark
+     * Evict records from the cache until the number of records is
+     * equal to or less than the cache capacity. Eviction is performed
+     * by repeatedly evicting all of the records with the smallest address.
      */
-    public void invalidateUpTo(long trimMark) {
-        log.debug("Invalidate sequencer cache. Trim mark: {}", trimMark);
-        int entries = 0;
-        int pqEntries = 0;
-        while (Address.isAddress(firstAddress()) && firstAddress() < trimMark) {
-            pqEntries += invalidateSmallestTxVersion();
-            entries++;
+    private void evict() {
+        while (conflictKeyMap.size() > capacity) {
+            evictSmallestTxVersion();
```

Comment:
I think it might be useful to mention that the cache releases memory via two mechanisms. One, eviction and two, overwriting. Imaging the case where you are continuously over writing the same conflict keys and the number of conflict parameters is less than capacity. This method will not evict anything, but the memory release happens in the calling method via overwriting.

Reply:
This has been reworked (only one data structure) after incorporating the use of a LinkedHashMap.